server/services/v1/key_generator.go

```diff
@@ -0,0 +1,131 @@
+package v1
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/internal"
+	"github.com/tigrisdata/tigris/keys"
+	"github.com/tigrisdata/tigris/lib/uuid"
+	"github.com/tigrisdata/tigris/schema"
+	"github.com/tigrisdata/tigris/server/metadata/encoding"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"github.com/tigrisdata/tigris/store/kv"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	// generatorSubspaceKey is used to store ids in storage so that we can guarantee uniqueness
+	generatorSubspaceKey = "generator"
+	// int32IdKey is the prefix after generator subspace to store int32 counters
+	int32IdKey = "int32_id"
+	// timestampIdKey is the prefix after generator to store timestamps
+	timestampIdKey = "timestamp_id"
+)
+
+type generator struct {
+	txMgr *transaction.Manager
+}
+
+func newGenerator(txMgr *transaction.Manager) *generator {
+	return &generator{
+		txMgr: txMgr,
+	}
+}
+
+// get returns generated id for the supported primary key fields.
+func (g *generator) get(ctx context.Context, table []byte, field *schema.Field) ([]byte, error) {
+	switch field.Type() {
+	case schema.StringType, schema.UUIDType:
+		return []byte(fmt.Sprintf(`"%s"`, uuid.NewUUIDAsString())), nil
+	case schema.ByteType:
+		return jsoniter.Marshal([]byte(uuid.NewUUIDAsString()))
+	case schema.DateTimeType, schema.Int64Type, schema.Int32Type:
+		return g.generateInTx(ctx, table, field)
+	}
+	return nil, api.Errorf(codes.InvalidArgument, "unsupported type found in auto-generator")
+}
+
+// generateInTx is used to generate an id in a transaction. This is mainly used where there is no easy way to guarantee
+// uniqueness so what we are doing is reserving this id in storage before returning to the caller so that only one
+// id is assigned to one caller.
+func (g *generator) generateInTx(ctx context.Context, table []byte, field *schema.Field) ([]byte, error) {
+	for {
+		tx, err := g.txMgr.StartTx(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var valueTS time.Time
+		var valueI64 int64
+		var valueI32 int32
+		switch field.Type() {
+		case schema.DateTimeType:
+			valueTS, _, err = g.generateTimestamp(ctx, tx, table)
+		case schema.Int64Type:
+			_, valueI64, err = g.generateTimestamp(ctx, tx, table)
```

Comment:
Int32 is sequential and Int64 is random?

Reply:
yeah to reduce the contention for int64, I am using timestamp nano.

Reply:
Can we just randomly genererate every type and regenerate if there is a duplicate key error when we insert the document?

Reply:
Optimistic approach avoid any synchronization of metadata.

Reply:
how would that work in case user is using Replace API?